pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1777,6 +1779,143 @@ public int reloadSegment(String tableNameWithType, String segmentName) {
+    Map<String, String> externalViewStateMap = externalView.getStateMap(segmentName);
+    List<String> partitions = Lists.newArrayList(segmentName);
+
+    // First, disable or reset partition
+    for (String instance : instanceSet) {
+      if (externalViewStateMap == null || SegmentStateModel.ERROR.equals(externalViewStateMap.get(instance))) {
+        LOGGER.info("Resetting partition: {} of table: {}", segmentName, tableNameWithType);
+        _helixAdmin.resetPartition(_helixClusterName, instance, tableNameWithType, partitions);
+      } else {
+        LOGGER.info("Disabling partition: {} of table: {}", segmentName, tableNameWithType);
+        _helixAdmin.enablePartition(false, _helixClusterName, instance, tableNameWithType, partitions);
+      }
+    }
+
+    // Wait for external view to stabilize
+    LOGGER.info("Waiting {} ms for external view to stabilize after disable/reset of partition: {} of table: {}",
+        externalViewWaitTimeMs, segmentName, tableNameWithType);
+    long startTime = System.currentTimeMillis();
+    Set<String> instancesToCheck = new HashSet<>(instanceSet);
+    while (!instancesToCheck.isEmpty() && System.currentTimeMillis() - startTime < externalViewWaitTimeMs) {
+      ExternalView newExternalView = getTableExternalView(tableNameWithType);
+      Preconditions
+          .checkState(newExternalView != null, "Could not find external view for table: %s", tableNameWithType);
+      Map<String, String> newExternalViewStateMap = newExternalView.getStateMap(segmentName);
+      if (newExternalViewStateMap == null) {
+        continue;
+      }
+      instancesToCheck.removeIf(instance -> SegmentStateModel.OFFLINE.equals(newExternalViewStateMap.get(instance)));
+    }
+    if (!instancesToCheck.isEmpty()) {
+      throw new IllegalStateException(String.format(
+          "Timed out waiting for external view to stabilize after disable/reset call. Skipping enable of partition: %s of table: %s",
+          segmentName, tableNameWithType));
+    }
+
+    // Enable partition
+    LOGGER.info("Enabling partition: {} of table: {}", segmentName, tableNameWithType);
+    for (String instance : instanceSet) {
+      _helixAdmin.enablePartition(true, _helixClusterName, instance, tableNameWithType, partitions);
+    }
+  }
+
+  /**
+   * Resets all segments of a table by disabling and then enabling the segments
+   */
+  public void resetAllSegments(String tableNameWithType, long externalViewWaitTimeMs) {
+    IdealState idealState = getTableIdealState(tableNameWithType);
+    Preconditions.checkState(idealState != null, "Could not find ideal state for table: %s", tableNameWithType);
+    ExternalView externalView = getTableExternalView(tableNameWithType);
+    Preconditions.checkState(externalView != null, "Could not find external view for table: %s", tableNameWithType);
+
+    Map<String, Set<String>> resetInstanceToPartitionsMap = new HashMap<>();
```

Comment:
suggest naming the variables with `segments` instead of `partritions`. We do have two other semantics of partitions that is already confusing (stream partitions, and partitioning of data in segment assignment)

Reply:
I want to keep it partitions because the calls being made are `resetPartition` and `enablePartition`

Reply:
Yes, when we make calls to helix we have to use partition. But within pinot we always refer to a segment as a segment . It will really reduce confusion if you keep the `segment` term when we refer to that.